src/model_DAforcing.py

```diff
@@ -175,6 +375,285 @@ def run(self, values: dict):
             #write flowveldepth file
             pass
 
+          
+def _time_stations_from_df(dataFrame, timeBase):
+
+    # get columns (date-time), extract as list at first
+    datesList = (dataFrame.columns).tolist()
+    # then subtract timebase
+    datesListSeconds = [int((d-timeBase).total_seconds()) for d in datesList]
+    nDates = len(datesListSeconds)
+    datesSecondsArray = np.array(datesListSeconds)
+
+    # get station IDs, as list
+    stations = (dataFrame).index.tolist()
+    nStations = len(stations)
+
+    # build station array (one giant 1D ndarray with ASCII encodings
+    # concatenated together), aided by an array of how many strings each
+    # station ID consists of - completely USGS, USACE, etc agnostic 
+    #
+    # define array defining string length for each station ID
+    stationStringLengthArray = np.empty(nStations, dtype=np.int64)
+    stationStringLengthArray[0] = len(stations[0])
+
+    # start big array with ID for first station
+    stationArray = np.empty(stationStringLengthArray[0], dtype=np.int64)
+
+    # init some aux. variables
+    firstIter = True
+    stationIndex = 0
+    # iterate through all stations
+    for station in stations:
+        # capture string length of present station
+        stationStringLengthArray[stationIndex] = len(station)
+        stationIndex += 1        
```

Comment:
I'll note here that BMI 2.0 doesn't explicitly allow passing arbitrary string type variables. This function may still be useful in the future with changes or extensions of BMI, but there will need to be some extra care in handling the string buffers.

Reply:
Just to clarify, none of the arrays contain strings (or characters) - sorry if the variable naming was misleading. Entries in dataframes are converted to integers by encoding into ASCII. The entire dataframe segment (e.g., column of station IDs) is converted into one single integer array, simply concatenating all the ASCII array representations of each station IDs together. Since we cannot always rely on each string (station ID) to have the same character length, a second array is created with the string lengths of the station IDs (or other entries such as filenames, etc). 

The specific array you flagged is actually said array of string lengths. 

So: the arrays as implemented are definitely BMI compliant.

Reply:
Hmm, I can see a possible path here...
The problem, I think, lies in this function in BMI 2.0:
https://bmi.readthedocs.io/en/stable/#get-var-itemsize
It requires any BMI variable (which may be an array) to say how many bytes a SINGLE element of that array contains (the entire array would have a total size given by https://bmi.readthedocs.io/en/stable/#get-var-nbytes).
So if you are ASCII encoding variable length strings into an array, you will need to pad each string to some maximum size.  Also should probably report the units of that variable as "ASCII" to indicate the encoding used.

We could extend the typically supported types to include "cstring"  in this way.

Unless the intent is to pass these as `char`, or unsigned short `int` arrays, which I think would work as is, but will require some extra protocol to decode.

Thanks for explaining a bit of this rational!

Reply:
To clarify further: each entry of the array is simply an ASCII code, so indeed, short unsigned int would suffice. So, a set of entries in a dataframe that are, e.g., station IDs are encoded as TWO integer arrays. So, e.g., the two following strings: 

['S1', '007'] are encoded as:

stringArray: 
        ([ASCII code of 'S', ASCII code of '1', ASCII code of '0', ASCII code of '0', ASCII code of '7']), 
or:   
        ([83, 49, 48, 48, 55 ])

stringLengthArray:
        ([Length of 'S1'], [Length of '007']) , 
or:    ([2, 3]). 

Both encoding and decoding are implemented based on that.

Reply:
Each item of the array is well defined, and can be represented as the same size integer. The point that we could use unsigned short integers is well taken, and we can definitely take a note for the next PR: will that also comply with any non-python components planned?

Reply:
I'm thinking "out loud" on this thread... Moving arbitrarily sized strings (character/byte arrays) via BMI isn't something heavily explored, and isn't currently supported in the ngen runtime, so thinking about how that would be done. A fixed size array with `'\0'` terminated strings could avoid the need for a separate length array, but that depends on what the `"S1"` semantics are in the context of the identifier.